interfaces/prompting/patterns/parse.go

```diff
@@ -0,0 +1,132 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package patterns
+
+import (
+	"errors"
+	"fmt"
+)
+
+func parse(tokens []token) (renderNode, error) {
+	tr := tokenReader{
+		tokens: tokens,
+	}
+	return parseSeq(&tr)
+}
+
+func parseSeq(tr *tokenReader) (renderNode, error) {
+	var s seq
+seqLoop:
+	for {
+		t := tr.peek()
+
+		switch t.tType {
+		case tokEOF:
+			break seqLoop
+		case tokBraceOpen:
+			inner, err := parseAlt(tr)
+			if err != nil {
+				return nil, err
+			}
+
+			s = append(s, inner)
+		case tokBraceClose:
+			if tr.depth > 0 {
+				break seqLoop
+			}
+			tr.token()
+			return nil, errors.New("unmatched '}' character")
+		case tokText:
+			tr.token()
+			s = append(s, literal(t.text))
+		case tokComma:
+			if tr.depth > 0 {
+				break seqLoop
+			}
+			tr.token() // discard, we get called in a loop
+			s = append(s, literal(","))
+		}
+	}
+
+	return s.optimize(), nil
+}
+
+func parseAlt(tr *tokenReader) (renderNode, error) {
+	var a alt
+
+	if t := tr.token(); t.tType != tokBraceOpen {
+		// Should not occur, caller should call parseAlt on peeking '{'
+		return nil, fmt.Errorf("internal error: expected '{' at start of alt, but got %v", t)
```

Comment:
maybe there are some errors that should be internal? I think I saw at least another comment about something that shouldn't occur

Reply:
Yes, there were some things which should never occur (e.g. a seq with length 0 after it's been parsed) which used to have checks which hid those errors, and I instead opted to remove the checks as they should be impossible for those errors to occur. But I think I'll add them back in so that if memory is corrupted or something in the code is changed, we'll get an error instead of a panic.